boa/src/syntax/ast/node/iteration/for_in_of_loop/mod.rs

```diff
@@ -8,68 +9,113 @@ use crate::{
 };
 use std::fmt;
 
+use crate::builtins::iterable::IteratorRecord;
 #[cfg(feature = "deser")]
 use serde::{Deserialize, Serialize};
 
 #[cfg_attr(feature = "deser", derive(Serialize, Deserialize))]
 #[derive(Clone, Debug, Trace, Finalize, PartialEq)]
-pub struct ForOfLoop {
+pub enum IterationKind {
+    Iterate,
+    Enumerate,
+}
+
+#[cfg_attr(feature = "deser", derive(Serialize, Deserialize))]
+#[derive(Clone, Debug, Trace, Finalize, PartialEq)]
+pub struct ForInOfLoop {
```

Comment:
I understand that the logic is very similar for `for...in` and `for...of`, but it seems that this requires some extra fields that will not be space efficient, right?

Reply:
My initial thought was also to add a new node, but then I read the spec and both `for..in` and `for..of` are described in the same sections. It could make things easier when adding `for await..in` and `for await..of` as they are also described in the same sections. Nevertheless, I understand this is less space efficient, and that we can add a new node for `for..in`.

Reply:
Yes, even if they are described in the same section, I think it's best if we add a different node for space efficiency reasons :) As long as we follow the spec in functionality, we don't need to have the code organised exactly in the same way.

Reply:
I have split `ForInOfLoop`.